File: shells/ios-template/WefterBridge/WefterPlugin.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

open class WefterPlugin {
    public let dispatcher: BridgeDispatcher

    #if canImport(UIKit)

    public weak var viewController: UIViewController?

    public required init(dispatcher: BridgeDispatcher, viewController: UIViewController?) {
        self.dispatcher = dispatcher
        self.viewController = viewController
    }
    #else
    public required init(dispatcher: BridgeDispatcher) {
        self.dispatcher = dispatcher
    }
    #endif

    public func resolve(_ callback: (Result<Any, Error>) -> Void, data: [String: Any] = [:]) {
        callback(.success(data))
    }

    public func reject(_ callback: (Result<Any, Error>) -> Void, code: String, message: String) {
        callback(.failure(WefterError(code: code, message: message)))
    }

    public func emit(_ hookName: String, _ data: [String: Any]) {
        dispatcher.emit(hookName, data)
    }
}

File: shells/android-template/app/src/main/java/dev/wefter/bridge/WefterPlugin.kt
package dev.wefter.bridge

import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import org.json.JSONObject

abstract class WefterPlugin(
        protected val context: Context,
        protected val dispatcher: BridgeDispatcher
) {
    protected fun resolve(callback: (Result<Any>) -> Unit, data: JSONObject = JSONObject()) =
            callback(Result.success(data))

    protected fun reject(callback: (Result<Any>) -> Unit, code: String, message: String) =
            callback(Result.failure(WefterError(code, message)))

    protected fun emit(hookName: String, data: JSONObject) = dispatcher.emit(hookName, data)

    protected fun hasPermission(permission: String): Boolean =
            ContextCompat.checkSelfPermission(context, permission) ==
                    PackageManager.PERMISSION_GRANTED

    protected fun requestPermission(
            activity: Activity,
            permission: String,
            onResult: (granted: Boolean) -> Unit
    ) = dispatcher.requestPermission(activity, permission, onResult)
}
